sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetField.scala

```diff
@@ -0,0 +1,199 @@
+        ArrayOrdinalGetField(child, fieldExpr)
+      case (_: MapType, _) =>
+        MapOrdinalGetField(child, fieldExpr)
+      case (otherType, _) =>
+        throw new AnalysisException(
+          "GetField is not valid on child of type " +
+            s"$otherType with fieldExpr of type ${fieldExpr.dataType}")
+    }
+  }
+
+  def unapply(g: GetField): Option[(Expression, Expression)] = {
+    g match {
+      case _: StructGetField => Some((g.child, null))
+      case o: OrdinalGetField => Some((o.child, o.ordinal))
+      case _ => None
+    }
+  }
+
+  /**
+   * find the ordinal of StructField, report error if no desired field or over one
+   * desired fields are found.
+   */
+  private def findField(fields: Array[StructField], fieldName: String, resolver: Resolver): Int = {
+    val checkField = (f: StructField) => resolver(f.name, fieldName)
+    val ordinal = fields.indexWhere(checkField)
+    if (ordinal == -1) {
+      throw new AnalysisException(
+        s"No such struct field $fieldName in ${fields.map(_.name).mkString(", ")}")
+    } else if (fields.indexWhere(checkField, ordinal + 1) != -1) {
+      throw new AnalysisException(
+        s"Ambiguous reference to fields ${fields.filter(checkField).mkString(", ")}")
+    } else {
+      ordinal
+    }
+  }
+}
+
+trait GetField extends UnaryExpression {
+  self: Product =>
+
+  type EvaluatedType = Any
+}
+
+abstract class StructGetField extends GetField {
+  self: Product =>
+
+  def field: StructField
+
+  override def foldable: Boolean = child.foldable
+  override def toString: String = s"$child.${field.name}"
+}
+
+abstract class OrdinalGetField extends GetField {
+  self: Product =>
+
+  def ordinal: Expression
+
+  /** `Null` is returned for invalid ordinals. */
+  override def nullable: Boolean = true
+  override def foldable: Boolean = child.foldable && ordinal.foldable
+  override def toString: String = s"$child[$ordinal]"
+  override def children: Seq[Expression] = child :: ordinal :: Nil
+
+  override def eval(input: Row): Any = {
+    val value = child.eval(input)
+    if (value == null) {
+      null
+    } else {
+      val o = ordinal.eval(input)
+      if (o == null) {
+        null
+      } else {
+        evalNotNull(value, o)
+      }
+    }
+  }
+
+  protected def evalNotNull(value: Any, ordinal: Any): Any
+}
+
+/**
+ * Returns the value of fields in the Struct `child`.
+ */
+case class SimpleStructGetField(child: Expression, field: StructField, ordinal: Int)
+  extends StructGetField {
+
+  override def dataType: DataType = field.dataType
+  override def nullable: Boolean = child.nullable || field.nullable
+
+  override def eval(input: Row): Any = {
+    val baseValue = child.eval(input).asInstanceOf[Row]
+    if (baseValue == null) null else baseValue(ordinal)
+  }
+}
+
+/**
+ * Returns the array of value of fields in the Array of Struct `child`.
+ */
+case class ArrayStructGetField(
```

Comment:
see https://issues.apache.org/jira/browse/SPARK-2096.
"arrayOfStruct.field1" means to access all values of "field1" in this array of structs and the returns those values as an array.